onnx/defs/schema.h

```diff
@@ -1136,74 +1136,127 @@ class OpSchemaRegistry final : public ISchemaRegistry {
+        if (schema_ver_map.count(ver)) {
+          if (fail_duplicate_schema) {
+            const auto& schema = schema_ver_map[ver];
+            std::stringstream err;
+            err << "Trying to register schema with name " << op_name << " (domain: " << op_domain << " version: " << ver
+                << ") from file " << op_schema.file() << " line " << op_schema.line()
+                << ", but it is already registered from file " << schema.file() << " line " << schema.line() << std::endl;
+            fail_schema(err.str());
+          }
           return;
         }
-        auto ver_range_map = DomainToVersionRange::Instance().Map();
-        auto ver_range_it = ver_range_map.find(op_domain);
-        if (ver_range_it == ver_range_map.end()) {
-          std::stringstream err;
-          err << "Trying to register schema with name " << op_name << " (domain: " << op_domain << " version: " << ver
-              << ") from file " << op_schema.file() << " line " << op_schema.line() << ", but its domain is not"
-              << " known by the checker." << std::endl;
-
-          fail_schema(err.str());
-        }
-        auto lower_bound_incl = ver_range_it->second.first;
-        auto upper_bound_incl = ver_range_it->second.second;
-        if (!(lower_bound_incl <= ver && upper_bound_incl >= ver)) {
-          std::stringstream err;
-          err << "Trying to register schema with name " << op_name << " (domain: " << op_domain << " version: " << ver
-              << ") from file " << op_schema.file() << " line " << op_schema.line() << ", but its version is not "
-              << "in the inclusive range [" << lower_bound_incl << ", " << upper_bound_incl
-              << "] (usually, this means you "
-              << "bumped the operator version but "
-              << "forgot to update the version range in DomainToVersionRange "
-              << "in onnx/defs/schema.h)." << std::endl;
-          fail_schema(err.str());
+
+        if (opset_version_to_load != 0) {
+          // Stops because the opset_version is higher than opset_version_to_load
+          if (ver > opset_version_to_load) return;
+
+          // Stops because a later version is registered within target opset version
+          if (!schema_ver_map.empty()) {
+            int max_registered_ver_le_target = GetMaxRegisteredVerWithinTarget(
+              schema_ver_map, opset_version_to_load);
+            if (max_registered_ver_le_target >= ver) return;
+          }
         }
 
-        m[op_name][op_domain].insert(std::pair<int, OpSchema&&>(ver, std::move(op_schema)));
+        CheckDomainAndVersionToRegister(op_schema, op_name, op_domain);
+        schema_ver_map.insert(std::pair<int, OpSchema&&>(ver, std::move(op_schema)));
       }
       ONNX_CATCH(const std::exception& e) {
         ONNX_HANDLE_EXCEPTION([&]() { std::cerr << "Schema error: " << e.what() << std::endl; });
       }
     }
+
+   private:
```

Comment:
Separated out the static utility codes to simplify the `OpSchemaRegisterOnce` logic above.

Reply:
Great! Thank you for refactoring.